File: fuse/python/fusepy/addresses.py
from fusepy._fusepy import *
from fusepy._fusepy import _FileOffset, _AddressFormat, _AddressLayout, _RelativeAddressLayout, _AddressMap, _RegionedAddressMap, _MipsEmbeddedLayout

class AddressLayout:
    @staticmethod
    def _make(layout):
        if isinstance(layout, _RelativeAddressLayout):
            return RelativeAddressLayout._make(layout)
        elif isinstance(layout, _MipsEmbeddedLayout):
            return MipsEmbeddedLayout._make(layout)
        elif isinstance(layout, _AddressLayout):
            return AddressLayout(layout)
        else:
            raise TypeError("expected an _AddressLayout or subclass")

    def __init__(self, layout):
        self._layout = layout

    def to_dict(self):
        return self._layout.to_dict()

class RelativeAddressLayout(AddressLayout):
    @staticmethod
    def _make(_layout):
        layout = RelativeAddressLayout.__new__(RelativeAddressLayout)
        layout._layout = _layout
        return layout

    def __init__(self, base, size, null_pointer=None, signedness=Signedness.UNSIGNED, endianness=Endianness.LITTLE):
        self._layout = _RelativeAddressLayout()
        self._layout.set_layout(size, signedness.value, endianness.value)
        if isinstance(base, int):
            self._layout.set_fixed_base_address(int(base))
        else:
            raise TypeError("expected an integer as base")
        if null_pointer is not None:
            self._layout.set_null_pointer(null_pointer[0], null_pointer[1])

class MipsEmbeddedLayout(AddressLayout):
    @staticmethod
    def _make(_layout):
        layout = MipsEmbeddedLayout.__new__(MipsEmbeddedLayout)
        layout._layout = _layout
        return layout

    def __init__(self, base, lo16, hi16):
        self._layout = _MipsEmbeddedLayout()
        self._layout.set_offsets(hi16, lo16)
        self._layout.set_base_address(base)

class AddressFormat:
    def __init__(self, format):
        if not isinstance(format, _AddressFormat):
            raise TypeError("expected an _AddressFormat")
        self._format = format

FileOffset = AddressFormat(_FileOffset)

class AddressMap:
    def __init__(self, map):
        if not isinstance(map, _AddressMap):
            raise TypeError("expected an _AddressMap")
        self._map = map

class RegionedAddressMap(AddressMap):
    def __init__(self, source_format, target_format):
        self._map = _RegionedAddressMap(source_format._format, target_format._format)

    def add_region(self, source, target, size):
        self._map.add_region(source, target, size)

File: python/src/kaizopy/addresses.cc
#include "addresses.h"
#include "binary.h"
#include <fuse/addresses/AbsoluteOffset.h>
#include <fuse/addresses/MipsLayout.h>
#include <fuse/addresses/RegionAddressMap.h>
#include <fuse/addresses/RelativeOffsetLayout.h>
#include <pybind11/stl.h>

using namespace kaizo::data;
namespace py = pybind11;

auto AddressLayout_encode(const AddressLayout& layout, const Address& address) -> py::list
{
    auto patches = layout.writeAddress(address);
    py::list list(patches.size());
    for (size_t i = 0; i < patches.size(); ++i)
    {
        list[i] = std::move(patches[i]);
    }
    return list;
}

void RelativeAddressLayout_set_layout(RelativeOffsetLayout& layout, const size_t size,
                                      const Signedness signedness, const Endianness endianness)
{
    IntegerLayout integerLayout;
    integerLayout.sizeInBytes = size;
    integerLayout.signedness = signedness;
    integerLayout.endianness = endianness;
    layout.setOffsetFormat(integerLayout);
}

auto RelativeAddressLayout_to_dict(const RelativeOffsetLayout& layout) -> py::dict
{
    py::dict dict;
    if (layout.hasNullPointer())
    {
        py::dict pyNullPointer;
        pyNullPointer["offset"] = layout.nullPointer().offset;
        pyNullPointer["address"] = layout.nullPointer().address;
        dict["null_pointer"] = pyNullPointer;
    }
    dict["base"] = layout.baseAddress();

    py::dict pyIntegerLayout;
    pyIntegerLayout["size"] = layout.offsetLayout().sizeInBytes;
    pyIntegerLayout["endianness"] = layout.offsetLayout().endianness;
    pyIntegerLayout["signedness"] = layout.offsetLayout().signedness;
    dict["layout"] = pyIntegerLayout;

    return dict;
}

static auto AddressMap_map_to_sources(const AddressMap& map, const Address address) -> py::list
{
    auto const sources = map.toSourceAddresses(address);
    py::list list(sources.size());
    for (size_t i = 0; i < sources.size(); ++i)
    {
        list[i] = sources[i];
    }
    return list;
}

//#################################################################################################

void registerFuseAddresses(py::module_& m)
{
    py::class_<Address>(m, "Address")
        .def("__str__", &Address::toString)
        .def("offset", &Address::applyOffset);
    py::class_<AddressFormat>(m, "_AddressFormat").def("from_int", &AddressFormat::fromInteger);
    m.add_object("_FileOffset", py::cast(static_cast<const AddressFormat*>(fileOffsetFormat()),
                                         py::return_value_policy::reference));

    py::class_<AddressMap>(m, "_AddressMap")
        .def("map_to_sources", &AddressMap_map_to_sources)
        .def("map_to_target", [](const AddressMap& map, const Address address) {
            return map.toTargetAddress(address);
        });
    py::class_<RegionAddressMap, AddressMap>(m, "_RegionedAddressMap")
        .def(py::init<const AddressFormat*, const AddressFormat*>())
        .def("add_region", &RegionAddressMap::map);

    py::class_<AddressLayout>(m, "_AddressLayout")
        .def("encode", &AddressLayout_encode)
        .def_property_readonly("id", &AddressLayout::getName);
    py::class_<RelativeOffsetLayout, AddressLayout>(m, "_RelativeAddressLayout")
        .def(py::init())
        .def("set_layout", &RelativeAddressLayout_set_layout)
        .def("set_null_pointer", &RelativeOffsetLayout::setNullPointer)
        .def("set_fixed_base_address", &RelativeOffsetLayout::setBaseAddress)
        .def("to_dict", &RelativeAddressLayout_to_dict);
    py::class_<MipsLayout, AddressLayout>(m, "_MipsEmbeddedLayout")
        .def(py::init())
        .def("set_base_address", &MipsLayout::setBaseAddress)
        .def("set_offsets", &MipsLayout::setOffsets);
}
